Stripe and orders table implementation

georgestam opened the pull request:
@vemv PR from issue #43 and #24. 
The workflow is as follows:
- Teacher input hours in their booking card (as per 9fd8bd20309c361a3be69edc58722fb64ffdf183)
- Whenever the teacher feels convenient, he/she clicks the button to create an invoice.
- Teacher input all the hours (in minutes) outstanding i.e. not paid yet), and create the order (old invoices are canceled by default). 
- An email is sent to the parent with a link to pay.
- Parent click the link, review the details and click the stripe button to pay the invoice.  

This feature works but I have the following comments:
- I cannot use your vemv engine because when `@order.charge!`is called,  I get:  `uninitialized constant Invoiceable::SpaceOrder`, as `invoiceable.rb`module calls to `space_order.rb`. Therefore I created a simple version of the action `charge` in  `order_controller.rb`
- I have added the `config.i18n.load_path += Dir[Rails.root.join('config/locales/**/*.yml').to_s]` in `application.rb`in order to have the mailers locales in `config/locales/en/mailers.yml`. However, it seems not to be working (I have located for now all the mailers translations in the main `en.yml`).  
- I still have to implement the testing.
- Rspec is now failing because mailers have to be moved  moved now within `mailers/action/..`

Commit:
Stripe and orders table implementation

georgestam commented on app/views/orders/checkout1.haml:
```diff
@@ -0,0 +1,20 @@
+    
+- content_for :bottom_js do    
+  = render partial: 'layouts/stripe_checkout', locals: {the_image_path: nil, the_button_selector: "##{the_button_id}", the_form_selector: "##{the_payment_form_id}", the_popup_title: the_popup_title, the_description: @order.journey.car.user.first_name, the_order: @order, the_hidden_input_selector: "##{the_hidden_input_id}"}
+
```
I tried to use a similar Kuina's checkout.haml. Do you see something incorrect? (I called `checkout1.haml` to disable it). When I click the pay button it does not do anything (and neither gives a `js` error in the console).

vemv replied:
You'll have to read and understand the code, also read the generated code (Chrome -> view source)

georgestam commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,136 @@
+  rescue Stripe::CardError => e
+    flash[:error] = e.message
+    redirect_to dashboard_path
+    
```
I can only introduce valid cards in stripe. For instance when I test stripe with an expired card I get: 
```Render and/or redirect were called multiple times in this action. Please note that you may only call render OR redirect, and at most once per action. Also note that neither redirect nor render terminate execution of the action, so if you want to exit an action after redirecting, you need to do something like "redirect_to(...) and return". ```
This is because I am unsure on how to locate the `rescue` action. I could do it as you do (with `if`/`else`, but i would like to do it like this in order to get the `e` error from stripe. Any ideas on how I I could achieve this?

vemv replied:
> For instance when I test stripe with an expired card I get: `Render and/or redirect were called multiple times in this action.`

This has nothing to do with Stripe. It's purely a Rails-level error. Focus on that first (double-render or double-redirect), and only then worry about Stripe details

georgestam commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,136 @@
+      end
+    else
+      flash.now[:error] = t 'orders.create.validation_error'
+      authorize(@order, :true?)
```
Could you please let me know what `@order, :true?` does?

georgestam commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,136 @@
+  
+  def new
+    @minutes = calculate_minutes
+    @order = Order.new
```
I would like to call the method `instantiate_order`from line 98 instead of doing this. How could I call `instantiate_order`(it needs an argument in line 98 but here it does not).

vemv replied:
Not worth doing

georgestam reviewed (commented):
@vemv Some additional queries on the code

georgestam commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,136 @@
+    #   flash[:error] = t 'stripe.payment_failed'
+    #   redirect_to @order.payment_link
+    # end
+    
```
This section is temporarily commented as I may be reused if I can bypass the error `uninitialized constant Invoiceable::SpaceOrder`

vemv commented on Gemfile:
```diff
@@ -41,6 +41,10 @@ gem 'friendly_id', '~> 5.1'
 
+gem "haml-rails", "~> 0.9"
+gem 'money-rails', '~>1'
+gem 'stripe', github: 'stripe/stripe-ruby', ref: 'cb841602fad553f8495fae05940fef975b85449e'
```
why are you using GH stripe version instead of normal?

georgestam replied:
https://github.com/stripe/stripe-ruby/ did not show the version. I guess that I have to check always the code versions here https://rubygems.org/gems/stripe/versions/1.57.1

vemv replied:
no, do it like this:

* install it without version
* see the version in the new Gemfile.lock
* update Gemfile
* run `bundle install` again (Gemfile.lock will be slightly different now)

vemv commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,136 @@
+  end 
+  
+  def create
+    @order.consumer_total = @order.minutes * @order.price_hour / 60
```
move this logic to the model

georgestam replied:
👍

vemv replied:
No need to notify me, once you update the code GH will mark/hide my comment as outdated

vemv commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,136 @@
+  
+  def create
+    @order.consumer_total = @order.minutes * @order.price_hour / 60
+    @order.state = 'not paid'
```
* states should be constants: `NOT_PAID = 'not_paid'`
* states shouldn't have spaces. Underscores instead

georgestam replied:
👍  I have added the following in `order.rb`
```NOT_PAID = 'not_paid'.freeze
PAID = 'paid'.freeze```

vemv commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,136 @@
+  protected
+  
+  def set_journey
+    @journey = Journey.find(params[:journey_id])
```
find!

vemv commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,136 @@
+  end
+  
+  def set_order
+    @order = Order.find(params[:id])
```
find!

vemv commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,136 @@
+  
+  def handle_errors
+    ensure_order
+    render 'oders/new'
```
"oders"

vemv commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,136 @@
+    end
+  end
+  
+  def calculate_minutes
```
move this logic to model

vemv commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,136 @@
+    
+  end
+  
+  def calculate_minutes_paid
```
move this logic to model

vemv commented on app/jobs/orders/creation/job.rb:
```diff
@@ -0,0 +1,15 @@
+    class Job < ApplicationJob
+      
+      def perform model_name, order_id
+        order = Order.find order_id
```
find!

vemv commented on app/mailers/application_mailer.rb:
```diff
@@ -1,5 +1,12 @@
+    subclass.layout 'mailer'
+  end
+  
+  # layout 'mailer'
```
remove

vemv commented on app/policies/order_policy.rb:
```diff
@@ -0,0 +1,28 @@
+    create_or_destroy?
+  end  
+  
+  def create_or_destroy?
```
create_or_destroy?` actually means in this case `create_or_destroy_or_checkout_or_charge`, see what I mean?

avoid method names which are "lies".

Rename it to `touch?`

vemv commented on app/views/orders/new.html.erb:
```diff
@@ -0,0 +1,38 @@
+
+            <%= simple_form_for [@journey, @order], url: journey_orders_path(@journey) do |f| %>
+              <%= f.error_notification %>
+              <%#= raise %>
```
remove

vemv commented on app/views/orders/new.html.erb:
```diff
@@ -0,0 +1,38 @@
+          <div class="dashboard-card">
+
+           <p class="text-center"> 
+              <%= t('.invoice_description', total_hours: (@minutes[:minutes_not_paid].to_f/60).round(1), total_min: @minutes[:minutes_not_paid] ) %> 
```
move `.to_f/60).round(1)` to model

georgestam replied:
I have `@minutes` as an instance variable (not as a class), therefore I created the helper method `convert_to_hours(hours)` in `order_controller.rb`.

vemv replied:
* sounds like the controller method would be better placed in model
* in any case, `.to_f/60).round(1)` does not belong to views. Knowledge of conversions/rounding belongs to models, or separate classes (informally known as 'presenters').

vemv commented on app/views/orders/new.html.erb:
```diff
@@ -0,0 +1,38 @@
+
+
+
+
```
Remove blank lines

vemv commented on config/application.rb:
```diff
@@ -25,12 +25,56 @@ class Application < Rails::Application
+      if development? || test?
+        'localhost'
+      else
+        nil
```
beware of ruining staging/production!

vemv commented on config/application.rb:
```diff
@@ -25,12 +25,56 @@ class Application < Rails::Application
+    # 
+    # config.action_controller.default_url_options ||= {}
+    # config.action_mailer.default_url_options ||= {}
+    # Rails.application.routes.default_url_options ||= {}
```
remove commented out code

Commit:
implement requested changes

Commit:
move previous mailers to 'mailers/class/action/...

georgestam commented:
@vemv I have implemented your changes in the first commit 1a1fac0da48d0bbe2407ef2e2e5369d735cb2e63
Then, I moved previous mailers to `mailers/class/action/...` in second commit 42059e211d9614985624d6367378884ca96c5bce

vemv commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,105 @@
+  end
+  
+  def new
+    @minutes = Order.calculate_minutes(@journey)
```
minutes of what?

I suggest renaming this variable to something more descriptive

imparted_minutes, billable_minutes, something like that (don't know the answer for this case)

vemv commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,105 @@
+  end 
+  
+  def create
+    @order.calculate_consumer_total_cents
```
rename this method to `calculate_consumer_total_cents!`

Use bang convention for methods which mutate state.

vemv replied:
Yes, mutating instance variables is a side-effect. `!` informs us of side effects

vemv commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,105 @@
+  
+  def create
+    @order.calculate_consumer_total_cents
+    @order.state = Order::NOT_PAID
```
NOT_PAID isn't quite a good name - it reads like "the customer didn't pay"

use something like pending, new, etc

vemv commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,105 @@
+    @order.calculate_consumer_total_cents
+    @order.state = Order::NOT_PAID
+    if @order.valid?
+      authorize(@order)
```
authorize(@order) is in both sides of the if/else, DRY it

vemv commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,105 @@
+      flash[:notice] = t '.success'
+      redirect_to dashboard_path
+    end 
+    rescue Stripe::CardError => e
```
indent `rescue` at the same level as `def`.

vemv commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,105 @@
+    
+    if @order.update(payment: charge.to_json, state: Order::PAID)
+      flash[:notice] = t '.success'
+      redirect_to dashboard_path
```
You don't seem to handle the case where `.update` returns false?

vemv commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,105 @@
+  end
+  
+  def ensure_order
+    if current_user
```
redundant condition - authentication is needed for all actions in this controller

lesson is - please understand what you are copy/pasting

vemv commented on app/mailers/journeys/creation/journey_mailer.rb:
```diff
@@ -0,0 +1,38 @@
+        @user = journey.user
+        @journey = journey
+        @car = @journey.car
+        
```
two consecutive newlines

vemv commented on app/mailers/journeys/creation/journey_mailer.rb:
```diff
@@ -0,0 +1,38 @@
+      end
+      
+      def select_from
+        if @user.domain == 'http://www.diverlang.com' || @user.domain == 'http://diverlang.com'
```
user.domain.to_s.include?('diverlang.com')

vemv commented on app/mailers/users/creation/user_mailer.rb:
```diff
@@ -0,0 +1,32 @@
+        @user = User.find(user_id)
+        # This will render a view in `app/views/user_mailer`!
+        
+        if @user.domain == 'http://www.diverlang.com' || @user.domain == 'http://diverlang.com'
```
create `User#diverlang?` method

vemv commented on app/views/profiles/dashboard/_journey.html.erb:
```diff
@@ -22,9 +22,11 @@
                 <li> <i class="fa fa-clock-o" aria-hidden="true"></i>  <%= t('.total_hours') %>: <%=  h[:total_hours] %> </li>
-                <li> <i class="fa fa-calendar" aria-hidden="true"></i>  <%= t('.hours_paid') %>: <%=  h[:hours_paid] %> </li>
-                <li> <i class="fa fa-calendar-times-o" aria-hidden="true"></i>  <%= t('.hours_not_paid') %>: <%= h[:hours_not_paid] %> </li>
+                <!-- <li> <i class="fa fa-calendar" aria-hidden="true"></i>  <%= t('.hours_paid') %>: <%=  h[:hours_paid] %> </li> -->
```
remove

vemv commented on app/views/profiles/dashboard/_journey.html.erb:
```diff
@@ -50,14 +52,18 @@
                       <%= link_to I18n.t('.Edit_booking', default: 'Edit Booking' ), edit_journey_path(journey), class: 'btn btn-primary custom-btn-width' %>
                       <%= link_to I18n.t('profiles.dashboard.journey.history'), journey_imparted_hours_path(journey), class: 'btn btn-success custom-btn-width' %>
                        <%#= link_to I18n.t('.Cancel_booking', default: 'Cancel Booking' ), journey_path(journey), data: { confirm: 'Are you sure?' }, method: :delete, class: 'btn btn-danger custom-btn-width' %>
+                    
```
two newlines

vemv commented on config/application.rb:
```diff
@@ -25,12 +25,50 @@ class Application < Rails::Application
+      if development? || test?
+        'localhost'
+      else
+        "http://www.helloveronica.com/"
```
`host`  should not include protocol, that's why the `protocol` method exists below.

vemv commented:
Feedback aside, I see no specs introduced. Try covering all functionality you introduce

Commit:
implement requested changes

georgestam reviewed (commented):
@vemv I have implemented the changes.

Commit:
move diverlang? method outside the private section

Commit:
add rspec testing for Order/s

georgestam commented on app/controllers/application_controller.rb:
```diff
@@ -1,5 +1,9 @@
+  include ApplicationControllerImpl
+  
+  public
 
```
Why do I have to add this `public` here? (took it from Kuina). It seems to me not required and `robocop` consider it an offense.

georgestam commented on spec/controllers/orders/create_spec.rb:
```diff
@@ -0,0 +1,55 @@
+          
+        it "works" do
+          # expect{the_action(reference.minutes)}.to change(Order, :count).by(1)
+        end
```
This example is currently commented out, as the `the action` for this example is not working since the params is missing `order`

![screen shot 2017-04-12 at 09 08 29](https://cloud.githubusercontent.com/assets/16335156/24955801/d97b3b04-1f7c-11e7-838c-71bef17d5be6.png)

The correct params are the following:

![screen shot 2017-04-12 at 09 08 44](https://cloud.githubusercontent.com/assets/16335156/24955981/8f9a0a28-1f7d-11e7-833a-5de185b492fb.png)

If I remove `require(:order)` from the following (within `orders_controller.rb`)
```
def order_params
    params.require(:order).permit(:minutes).merge(journey_id: @journey.id, price_hour: @journey.car.price_hour)
end
```

it works, but it is not correct. How can I modify `the action`to include the order param?

georgestam reviewed (commented):
@vemv I have implemented the specs for this feature. I have included 2 new queries in the PR. 

Also, I may recall incorrectly that you mentioned that tests should follow the structure describe/context/it without nesting (e.g. describe/describe/context/describe...). I found that `charge_spec.rb` had some nesting and followed the same structure. Please let me know your comments.

vemv commented on app/controllers/orders_controller.rb:
```diff
@@ -0,0 +1,92 @@
+  
+  def charge
+    
+    raise "already charged" if @order.charged?
```
Showing HTTP 500 errors to the user is never friendly

* redirect and `return` instead of raise
* Also report the issue to raygun and Rails.logger - as (I imagine) this line should never happen
* See `SafeLogger` (in my engine) for a sample class which does both things. You can implement a similar one.

vemv commented on app/mailers/journeys/creation/journey_mailer.rb:
```diff
@@ -0,0 +1,37 @@
+        @journey = journey
+        @car = @journey.car
+        
+        from = select_from
```
* rename `select_from` to `from`
* don't create local variable, just use a single method call: `mail(from: from, to: recipient, subject: 'Booking confirmation')`

vemv commented on app/models/order.rb:
```diff
@@ -0,0 +1,77 @@
+  
+  validates :state, inclusion: [PENDING_PAYMENT, PAID]
+  validates :price_hour, presence: true
+  validates :minutes, :numericality => { :greater_than_or_equal_to => 1 }, :presence => true
```
install https://atom.io/packages/ruby-syntax-replacer and update these

vemv commented on config/application.rb:
```diff
@@ -25,12 +25,50 @@ class Application < Rails::Application
+      if development? || test?
+        'localhost'
+      else
+        "www.helloveronica.com"
```
must it be diverlang.com in any case?

vemv commented on spec/controllers/orders/charge_spec.rb:
```diff
@@ -0,0 +1,138 @@
+          let!(:journey){ FactoryGirl.create :journey, user: user }
+          let!(:order){ FactoryGirl.create :order, :paid, journey: journey }
+
+          it "raise and exception" do
```
grammar

vemv commented on spec/controllers/orders/charge_spec.rb:
```diff
@@ -0,0 +1,138 @@
+
+        end
+
+        context "attempting to pay an order I did not booked with a teacher" do
```
grammar

important if you want to deliver code professionally :)

vemv reviewed (commented):
We're almost there!

Commit:
remove raise already_charged in order to avoid a 500 error but login the error

Commit:
implement requested changes

Commit:
fix order create_spec with correct params

Commit:
remove git traking coverage folder

georgestam reviewed (commented):
@vemv I have implemented the changes

vemv commented on app/controllers/application_controller.rb:
```diff
@@ -1,5 +1,8 @@
   protect_from_forgery with: :exception
+  
+  include ApplicationControllerImpl
+  # include SafeLoggerRaygun
```
Don't introduce commented out code. Please eliminate this habit entirely, consider it an essential step for becoming a pro

vemv reviewed (changes requested):
Build is red

Develop a sensibility around not passing PRs in red, will be a great habit

Once in green merge w/o asking!

Commit:
add app/assets/stylesheets/components/_avatars.scss again to git, as they were removed (somehow)

Commit:
remove commented out code used for testing

georgestam reviewed (commented):
@vemv done.

vemv commented:
Good one, happy we made it!

Maybe next time you can try smaller PRs. Preferrably self-contained (code should make sense / be useful in the context of that PR alone), but not 100% necessarily. Main points:

* have early feedback
* give me PRs I can handle

As I view it most important think is to **keep growing**. Never (or rarely) make the same mistakes twice!

* Cover with tests everything you make, make that part of your workflow (and not an afterthought)
* Self-code review each set of staged changes before commiting, avoiding silly mistakes

Overall feeling satisfied with quality of the final code, but try to achieve the same results more quickly.

georgestam commented:
Many thanks @vemv for the feedback. I am learning a lot with your help.

Yes, sometimes with the rush to push code I do stupid mistakes....

Hopefully now, with fewer holidays, we would be able to close PR's quicker!
